optaplanner-examples/src/main/java/org/optaplanner/examples/coachshuttlegathering/optional/score/CoachShuttleGatheringConstraintProvider.java

```diff
@@ -0,0 +1,124 @@
+                shuttleDestinationIsCoachOrHub(constraintFactory),
+                shuttleSetupCost(constraintFactory),
+                distanceFromPrevious(constraintFactory),
+                distanceBusStopToBusDestination(constraintFactory),
+                distanceCoachDirectlyToDestination(constraintFactory)
+        };
+    }
+
+    // ############################################################################
+    // Hard constraints
+    // ############################################################################
+
+    private Constraint coachStopLimit(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BusStop.class)
+                .filter(busStop -> busStop.getBus() instanceof Coach)
+                .groupBy(busStop -> (Coach) busStop.getBus(), count())
+                .filter((coach, stopCount) -> stopCount > coach.getStopLimit())
+                .penalizeLong("coachStopLimit", HardSoftLongScore.ofUninitialized(0, 1000000, 0),
+                        (coach, stopCount) -> stopCount - coach.getStopLimit());
+    }
+
+    private Constraint shuttleCapacity(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BusStop.class)
+                .filter(busStop -> busStop.getBus() instanceof Shuttle)
+                .groupBy(BusStop::getBus, sum(BusStop::getPassengerQuantity))
+                .filter((shuttle, requiredCapacity) -> requiredCapacity > shuttle.getCapacity())
+                .penalizeLong("shuttleCapacity", HardSoftLongScore.ofUninitialized(0, 1000, 0),
+                        (shuttle, requiredCapacity) -> requiredCapacity - shuttle.getCapacity());
+    }
+
+    // TODO: Implement this CS
+//    private Constraint coachCapacity(ConstraintFactory constraintFactory) {
+//        BiConstraintStream<Coach, Integer> coachPassengerQuantity = constraintFactory.from(BusStop.class)
+//                .filter(busStop -> busStop.getBus() instanceof Coach)
+//                .groupBy(busStop -> (Coach) busStop.getBus(), sum(BusStop::getPassengerQuantity));
+//    }
+
+    private Constraint transportTime(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BusStop.class)
+                .filter(busStop -> busStop.getTransportTimeToHub() != null && busStop.getTransportTimeRemainder() < 0)
+                .penalizeLong("transportTime", HardSoftLongScore.ONE_HARD, busStop -> -1 * busStop.getTransportTimeRemainder());
+    }
+
+    private Constraint shuttleDestinationIsCoachOrHub(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shuttle.class)
+                .filter(shuttle -> shuttle.getDestination() != null)
+                .join(StopOrHub.class, equal(Shuttle::getDestination, stopOrHub -> stopOrHub))
+                .filter((shuttle, stopOrHub) -> !stopOrHub.isVisitedByCoach())
+                .penalize("shuttleDestinationIsCoachOrHub", HardSoftLongScore.ofUninitialized(0, 1000000000, 0));
```

Comment:
For these very large numbers, I suggest you use `_` to separate thousands. `1000000000` becomes `1_000_000_000`.